charmhelpers/contrib/openstack/context.py

```diff
@@ -1779,3 +1779,24 @@ def __call__(self):
                     ctxt['memcache_server_formatted'],
                     ctxt['memcache_port'])
         return ctxt
+
+
+class ExtraPolicyContext(OSContextGenerator):
+    """A context to render extra policy for a service"""
+
+    def __init__(self, policyd):
+        """
+        @param policyd: an absolute path to a policy directory which is
+        configured in policy_dirs oslo.policy option.
+        """
+        self.policyd = policyd
+        # make it if it's not there
+        mkdir(policyd)
```

Comment:
Let's not put `mkdir(...)` here.  Despite the fact I hate that these contexts have side-effects (rather than just being contexts), this ought probably to be in the `__call__()` as now `mkdir` would need to be mocked *before* the file that contains the call to instantiate this context is imported.  Obviously, it could be in the form `ExtraPolicyContext()()`, but let's at least have a chance.

Reply:
I hate this behavior too but I could not find any other place that would make sense for multiple classic charms.

I can move this to __call__().

Reply:
I wasn't getting at you :)  ... it's a common pattern in the charms and tricky to fix - however, in the `__call__` at least we can refactor charms to move the call to `__call__()` out of the load-time code (eventually). Thanks for fixing!

Reply:
No negativity implied :^)

Just wanted to say that I had a similar impression about contexts being more than contexts which is not obvious when such a change is needed.